connectors/db/vector.py

```diff
@@ -189,52 +260,99 @@ def init_vector_store(self):
-            if line.startswith("# "):
-                # we found a title header, add it to metadata
-                metadata["title"] = line.lstrip("# ").strip()
-                break
-        else:
-            metadata["title"] = metadata["full_path"]
 
-        text_splitter = RecursiveCharacterTextSplitter(
-            chunk_size=self.vector_chunk_size,
-            chunk_overlap=self.vector_chunk_overlap,
-            separators=TXT_SEPARATORS,  # todo: split on more doc types
-        )
+    def has_markdown_headers(self, text):
+        """Checks if a document contains markdown headers."""
+        return bool(re.search(r"^#{1,6} ", text, re.MULTILINE))
+
+    def remove_tables_of_contents(self, chunks):
+        """Removes MKDocs-style TOCs and footers, even when containing Markdown links."""
+        TOC_PREFIXES = ["Skip to content", "Table of Contents"]
+        FOOTER_PATTERN = re.compile(r"Made with .*Material for MkDocs", re.IGNORECASE)
+
+        cleaned_chunks = []
+        for chunk in chunks:
+            stripped_chunk = chunk.strip()
+            
+            # Skip TOC-like chunks
+            if any(stripped_chunk.startswith(prefix) for prefix in TOC_PREFIXES):
+                continue
+            
+            # Skip footers with Markdown links
+            if FOOTER_PATTERN.search(stripped_chunk):
+                continue
+            
+            # Passed both checks, keep it
+            cleaned_chunks.append(chunk)
+
+        return cleaned_chunks
+
+    def split_to_document_chunks(self, text, metadata, max_chunk_size=4000):
+        """Uses markdown-aware chunking and drops any chunk over a hard size limit."""
+
+        # Use markdown-aware text splitting
+        if self.has_markdown_headers(text):
+            text_splitter = MarkdownHeaderTextSplitter(
+                headers_to_split_on=[
+                    ("#", "H1"),
+                    ("##", "H2"),
+                    ("###", "H3"),
+                    ("####", "H4"),
+                    ("#####", "H5"),
+                    ("######", "H6"),
+                ]
+            )
+
+            # MarkdownHeaderTextSplitter returns Documents, extract text first
+            markdown_chunks = text_splitter.split_text(text)
+            chunks = [doc.page_content for doc in markdown_chunks]  # Convert to raw text
+
+        else:
+            text_splitter = RecursiveCharacterTextSplitter(
+                chunk_size=2000,  # Keep chunks readable
+                chunk_overlap=200,  # Small overlap for context
+                separators=["\n\n", ". ", "? ", "! "]
+            )
+            chunks = text_splitter.split_text(text)
 
-        chunks = text_splitter.split_text(text_to_split)
         chunks = self.combine_small_chunks(chunks)
+        
+        chunks = self.remove_tables_of_contents(chunks)
 
+        # Apply a Hard Cutoff for Chunk Size
+        filtered_chunks = [chunk for chunk in chunks if len(chunk) <= max_chunk_size]
```

Comment:
In theory if we deal with setting a limit in the "combine_small_chunks" function we will not need this correct? It seems like we accidentally drop content with this hard limit?

Reply:
Good call. I forgot to remove that. Pushed up a change.